include/AST/Types/EnumTy.hpp

```diff
@@ -2,60 +2,111 @@
 #define GLU_AST_TYPES_ENUMTY_HPP
 
 #include "Basic/SourceLocation.hpp"
-#include "TypeBase.hpp"
+#include "Types/TypeBase.hpp"
 
+#include <cassert>
 #include <llvm/ADT/APInt.h>
-#include <llvm/ADT/STLExtras.h>
-#include <llvm/ADT/SmallVector.h>
+#include <llvm/ADT/ArrayRef.h>
 #include <llvm/ADT/StringRef.h>
+#include <llvm/Support/Allocator.h>
+#include <llvm/Support/TrailingObjects.h>
 #include <string>
 
+namespace glu {
+template <typename T> class InternedMemoryArena;
+}
+
 namespace glu::types {
 
-/// @brief EnumTy is a class that represents enumerations declared in code.
-class EnumTy : public TypeBase {
+struct Case {
+    llvm::StringRef name;
+    llvm::APInt value;
+
+    Case(llvm::StringRef const &n, llvm::APInt const &v) : name(n), value(v) { }
+};
+
+class EnumTy final : public TypeBase,
+                     private llvm::TrailingObjects<EnumTy, Case> {
 public:
-    struct Case {
-        std::string name;
-        llvm::APInt value;
-    };
+    friend llvm::TrailingObjects<EnumTy, Case>;
+    friend class InternedMemoryArena<TypeBase>;
 
 private:
-    std::string _name;
-    llvm::SmallVector<Case> _cases;
+    llvm::StringRef _name;
+    unsigned _numCases;
     SourceLocation _definitionLocation;
 
-public:
-    /// @brief Constructor for the EnumTy class.
+    // Method required by llvm::TrailingObjects to determine the number
+    // of trailing objects.
+    size_t
+        numTrailingObjects(typename llvm::TrailingObjects<EnumTy, Case>::OverloadToken<Case>) const
+    {
+        return _numCases;
+    }
+
+    EnumTy(std::string name, unsigned numCases, SourceLocation loc)
+        : TypeBase(TypeKind::EnumTyKind)
+        , _name(std::move(name))
+        , _numCases(numCases)
+        , _definitionLocation(loc)
+    {
+    }
+
     EnumTy(
-        std::string name, llvm::SmallVector<Case> cases,
-        SourceLocation definitionLocation
+        std::string const &name, llvm::ArrayRef<Case> const &cases,
```

Comment:
take least specific types. A reference to a reference is useless
```suggestion
        llvm::StringRef name, llvm::ArrayRef<Case> cases,
```